### PR TITLE
Handle partition key and hidden columns in wrapper in Hive

### PR DESCRIPTION
**Pull request by geraint0923**

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/ColumnarBinaryHiveRecordCursor.java [hunk 1]
```diff
@@ -93,32 +79,27 @@
...
 class ColumnarBinaryHiveRecordCursor<K>
-        extends HiveRecordCursor
+        implements RecordCursor
 {
     private final RecordReader<K, BytesRefArrayWritable> recordReader;
     private final K key;
     private final BytesRefArrayWritable value;
 
     @SuppressWarnings("FieldCanBeLocal") // include names for debugging
-    private final String[] names;
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/ColumnarBinaryHiveRecordCursor.java]: Remove `@SuppressWarnings` corresponding to this field declaration as well

File: presto-hive/src/main/java/com/facebook/presto/hive/ColumnarBinaryHiveRecordCursor.java [hunk 2]
```diff
@@ -188,76 +166,13 @@ public ColumnarBinaryHiveRecordCursor(RecordReader<K, BytesRefArrayWritable> rec
 
         for (int i = 0; i < columns.size(); i++) {
             HiveColumnHandle column = columns.get(i);
+            checkState(column.getColumnType() == REGULAR, "column type is partition key or hidden");
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/ColumnarBinaryHiveRecordCursor.java]: `column type must be regular`

Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/ColumnarBinaryHiveRecordCursor.java]: Do the same for other instances.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java [hunk 3]
```diff
@@ -104,16 +109,17 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
...
             List<Type> columnTypes = ImmutableList.copyOf(transform(hiveColumns, input -> typeManager.getType(input.getTypeSignature())));
-            return new RecordPageSource(columnTypes, recordCursor);
+            return new RecordPageSource(columnTypes, hiveRecordCursor);
         }
 
         throw new RuntimeException("Could not find a file reader for split " + hiveSplit);
     }
 
-    protected HiveRecordCursor getHiveRecordCursor(
+    protected RecordCursor getHiveRecordCursor(
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java]: Rename this method `getRegularColumnRecordCursor`?

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java [hunk 4]
```diff
@@ -145,4 +151,22 @@ protected HiveRecordCursor getHiveRecordCursor(
         }
         return null;
     }
+
+    public static List<HiveColumnHandle> buildDataColumns(List<HiveColumnHandle> columns, int[] fieldMapping)
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java]: We don't pass arrays as out parameter.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveRecordCursor.java [hunk 5]
```diff
@@ -13,38 +13,270 @@
...
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static com.facebook.presto.spi.type.TinyintType.TINYINT;
+import static com.facebook.presto.spi.type.Varchars.isVarcharType;
+import static com.google.common.collect.Maps.uniqueIndex;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.util.Objects.requireNonNull;
 
-public abstract class HiveRecordCursor
+public final class HiveRecordCursor
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveRecordCursor.java]: We don't defensively declare classes as `final`

File: presto-hive/src/main/java/com/facebook/presto/hive/RecordCursorProvider.java [hunk 6]
```diff
@@ -24,9 +25,9 @@
 import java.util.Optional;
 import java.util.Properties;
 
-public interface HiveRecordCursorProvider
+public interface RecordCursorProvider
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordCursorProvider.java]: I think it makes sense to continue call this `HiveRecordCursorProvider` given this interface is very specifically for Hive.

File: presto-spi/src/main/java/com/facebook/presto/spi/RecordCursor.java [hunk 7]
```diff
@@ -18,14 +18,19 @@
...
+
 public interface RecordCursor
         extends Closeable
 {
     long getTotalBytes();
 
     long getCompletedBytes();
 
-    long getReadTimeNanos();
+    default long getReadTimeNanos()
```
Comment (haozhun) [inline presto-spi/src/main/java/com/facebook/presto/spi/RecordCursor.java]: You don't need this. Just implement it in all the RecordCursors in Hive.

File: presto-spi/src/main/java/com/facebook/presto/spi/RecordCursor.java [hunk 8]
```diff
@@ -51,4 +56,18 @@ default long getSystemMemoryUsage()
 
     @Override
     void close();
+
+    default void closeWithSuppression(Throwable throwable)
```
Comment (haozhun) [inline presto-spi/src/main/java/com/facebook/presto/spi/RecordCursor.java]: Make this a util method in Hive plugin

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveRecordCursor.java [hunk 9]
```diff
@@ -13,38 +13,270 @@
...
+        return delegate.getReadTimeNanos();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return delegate.getSystemMemoryUsage();
+    }
+
+    public RecordCursor getRecordCursor()
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveRecordCursor.java]: `getRegularColumnsRecordCursor`

`@VisibleForTesting`

**Comment by haozhun** [comment]
I reviewed first and last commit, and looked at second commit briefly. Some of my comments for the first commit also applies to the second one. Please address them.

**Comment by haozhun** [comment]
I looked at the partition column block creation code in the second commit. Those look reasonable. Everything else in the second commit should be pretty routine and similar to the first commit. I'll do a pass for all the commits once you address my current comments.

**Comment by geraint0923** [comment]
Addressed the comments.

**Comment by geraint0923** [comment]
Addressed the comments.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java [hunk 10]
```diff
@@ -100,49 +136,117 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
...
                     typeManager);
+
             if (cursor.isPresent()) {
-                return cursor.get();
+                HiveRecordCursor hiveRecordCursor = new HiveRecordCursor(
+                        columnMappings,
+                        hiveStorageTimeZone,
+                        typeManager,
+                        cursor.get());
+                List<Type> columnTypes = ImmutableList.copyOf(transform(hiveColumns, input -> typeManager.getType(input.getTypeSignature())));
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java]: We don't use `transform` any more. Rewrite using streaming API. Don't worry about Immutable.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java [hunk 11]
```diff
@@ -69,22 +77,50 @@ public HivePageSourceProvider(
     @Override
     public ConnectorPageSource createPageSource(ConnectorTransactionHandle transaction, ConnectorSession session, ConnectorSplit split, List<ColumnHandle> columns)
     {
-        HiveSplit hiveSplit = checkType(split, HiveSplit.class, "split");
-
-        String clientId = hiveSplit.getClientId();
+        List<HiveColumnHandle> hiveColumns = ImmutableList.copyOf(transform(columns, HiveColumnHandle::toHiveColumnHandle));
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java]: We don't use `transform` any more. Rewrite using streaming API. Don't worry about Immutable.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java [hunk 12]
```diff
@@ -100,49 +136,117 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
...
+                        cursor.get());
+                List<Type> columnTypes = ImmutableList.copyOf(transform(hiveColumns, input -> typeManager.getType(input.getTypeSignature())));
+                return Optional.of(new RecordPageSource(columnTypes, hiveRecordCursor));
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public static class ColumnMapping
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java]: You abstracted this really well.

**Pull request merged**